tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1439 @@
+  /// @param start true if starting position desired otherwise false
+  /// @return The tag starting position, otherwise the position immediately after the tag
+  Result<std::size_t> FindTag(std::string_view data, std::string_view tag,
+                              std::size_t pos=0, bool start=true) {
+    if(pos=data.find(tag, pos); pos != std::string_view::npos) {
+      return start ? pos :  pos + tag.length();
+    }
+    return absl::NotFoundError(
+      absl::StrCat("Malformed List Response XML: can't find ", tag, " in ", data));
+  }
+
+  /// @brief Get tag contents within the supplied XML payload
+  /// @param data XML payload
+  /// @param open_tag Opening tag
+  /// @param close_tag Closing tag
+  /// @param pos Initial searching position. Updated with position immediately
+  ///            after the closing tag.
+  /// @return Tag contents
+  Result<std::string_view> GetTag(std::string_view data,
+                                  std::string_view open_tag,
+                                  std::string_view close_tag,
+                                  std::size_t * pos) {
+
+    TENSORSTORE_ASSIGN_OR_RETURN(auto tagstart, FindTag(data, open_tag, *pos, false));
+    TENSORSTORE_ASSIGN_OR_RETURN(auto tagend, FindTag(data, close_tag, tagstart, true));
+    *pos = tagend + close_tag.size();
+    return data.substr(tagstart, tagend - tagstart);
+  }
+
+
+  absl::Status OnResponseImpl(const Result<HttpResponse>& response) {
+    if (is_cancelled()) {
+      return absl::CancelledError();
+    }
+    ABSL_LOG_IF(INFO, TENSORSTORE_INTERNAL_S3_LOG_RESPONSES && response.ok())
+        << "List " << *response;
+
+    absl::Status status =
+        response.ok() ? HttpResponseCodeToStatus(*response) : response.status();
+    if (!status.ok() && IsRetriable(status)) {
+      return owner_->BackoffForAttemptAsync(status, attempt_++, this);
+    }
+
+    auto cord = response->payload;
+    auto payload = cord.Flatten();
+    auto kListBucketOpenTag = "<ListBucketResult xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">";
+    TENSORSTORE_ASSIGN_OR_RETURN(auto start_pos, FindTag(payload, kListBucketOpenTag, 0, false));
+    std::size_t pos = start_pos;
+    TENSORSTORE_ASSIGN_OR_RETURN(auto key_count_tag, GetTag(payload, "<KeyCount>", "<", &pos));
+    unsigned int keycount = 0;
+    if(!absl::SimpleAtoi(key_count_tag, &keycount)) {
+      return absl::InvalidArgumentError(absl::StrCat("Malformed KeyCount ", key_count_tag));
+    }
+
+    for(unsigned int k=0; k < keycount; ++k) {
+      if (is_cancelled()) {
+        return absl::CancelledError();
+      }
+      TENSORSTORE_ASSIGN_OR_RETURN(pos, FindTag(payload, "<Contents>", pos, false));
+      TENSORSTORE_ASSIGN_OR_RETURN(auto key_tag, GetTag(payload, "<Key>", "<", &pos));
+
+      if(!options_.range.empty() && tensorstore::Contains(options_.range, key_tag)) {
+        if (options_.strip_prefix_length) {
+          key_tag = key_tag.substr(options_.strip_prefix_length);
```

Comment:
check that length is not less than `options_.strip_prefix_length`

Reply:
done https://github.com/google/tensorstore/pull/91/commits/0e63d3bab83862b0abf4f968e1c11df8ccc1c74f